client/blocks/signup-form/index.jsx

```diff
@@ -163,6 +163,14 @@ class SignupForm extends Component {
 		this.maybeRedirectToSocialConnect();
 	}
 
+	UNSAFE_componentWillReceiveProps( nextProps ) {
+		if ( this.props.step?.status === 'processing' && nextProps.step?.status === 'invalid' ) {
```

Comment:
I'm not 100% confident these statuses proceed in a linear manner, but this is what appeared to happen in my testing:

<img width="306" alt="image" src="https://user-images.githubusercontent.com/36432/199788578-c6d80ecf-a775-49f3-bafd-efe624d3ba25.png">

```
diff --git a/client/blocks/signup-form/index.jsx b/client/blocks/signup-form/index.jsx
index 1e30e4649d..f70c7774f4 100644
--- a/client/blocks/signup-form/index.jsx
+++ b/client/blocks/signup-form/index.jsx
@@ -164,6 +164,7 @@ class SignupForm extends Component {
 	}
 
 	UNSAFE_componentWillReceiveProps( nextProps ) {
+		console.log( this.props.step?.status, nextProps.step?.status );
 		if ( this.props.step?.status === 'processing' && nextProps.step?.status === 'invalid' ) {
 			this.setState( {
 				submitting: false,

```

Reply:
Is it necessary to check if the current status is processing? We know for sure that the user is not `submitting` if the status is invalid.

Is using [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) a good idea in this case? Example:

```ts
static getDerivedStateFromProps( props ) {
    if( props.step?.status === 'invalid' ) {
      return {
        submitting: false
      }
    }

    return null
}
```

Reply:
> Is it necessary to check if the current status is processing? We know for sure that the user is not `submitting` if the status is invalid.

@zaguiini I don't think it's _necessary_. But, it would limit the context in which this change is applied, which makes me feel more confident about not causing some other bug.

> Is using [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) a good idea in this case? Example:

I considered it after reviewing the docs. However, I determined `UNSAFE_componentWillReceiveProps` to be a more "correct" fix because we shouldn't be managing state in this component anyway. `UNSAFE_componentWillReceiveProps` more clearly calls out that this is problematic.

> My main question is: why do we need to imperatively mutate the state here? Why? What kind of blind case are we hitting that isn't covered by our state management yet? I think answering this question will likely lead to a better fix than this attempt at patching the existing flow.

My concern with a "better fix" is that it will likely touch more code paths, which then leads to greater likelihood of breakage. I don't think the risks associated with refactoring are worth it, at the moment.

Reply:
>because we shouldn't be managing state in this component anyway.

Not really sure about this statement in specific. This component has different pieces of state that are, indeed, derived from the props, so it looks like the perfect match 🤔  Also, it will skip re-rendering (as it happens when you call `setState` inside `componentWillReceiveProps` so I do think it's a better option, and the lifecycle is going to be used as it's meant to be so it doesn't look inadequate at all.

>I don't think the risks associated with refactoring are worth it, at the moment.

That's also true. Let's ship this fix first, with your proposal or mine, then properly fix the issue as a follow-up.

Reply:
> then properly fix the issue as a follow-up.

These things have a tendency to be left the way they are for quite a long time, so I'd personally go with the `getDerivedStateFromProps` solution. It isn't quite as surgical but it's worked perfectly in my testing and doesn't add yet more `UNSAFE_` methods to clean up later.

Reply:
0@zaguiini @p-jackson Looking at this further, one side effect of `getDerivedStateFromProps()` is that the form would stay in a `submitting=false` state for several cycles until the step status changed again (as it will continue to remain `invalid` while the user completes the form).

I'd like to keep `UNSAFE_componentWillReceiveProps` because it's really the _transition_ from `processing` to `invalid` that we care about. When this event happens, we want to set `submitting=false` and be done with it. `getDerivedStateFromProps` doesn't support explicitly hooking into that transition.